### PR TITLE
Fix #91: audit(reviews): remediate PRs merged before review completion

### LINKED ISSUE #91: audit(reviews): remediate PRs merged before review completion
## Summary
Audit recent merged PRs in `viamin/agent-harness` that appear to have been merged before review was actually complete, and fix forward any real defects.

Auto-merge has been manually disabled for now. This issue tracks the cleanup work required before re-enabling it.

## Strong likely-premature merges
These PRs still had unresolved review threads at merge time:
- #81
- #55
- #57
- #42
- #16

## Additional weaker suspect cases
These did not show unresolved threads in the audit, but they do appear to have newer commits than the last recorded review:
- #83
- #80
- #1

### P1 first
- #55
- #81
- #57
- #42

### P2 next
- #16
- #83
- #80
- #1

## Expected output
- Review each PR's unresolved findings.
- Determine whether the finding is real, obsolete, or already fixed elsewhere.
- Open focused fix-forward PRs/issues for real defects.
- Record a disposition for every audited PR.

## Exit criteria
- Every PR in this audit has a recorded disposition.
- P1 defects have landed fixes.
- Remaining P2 findings are either fixed or explicitly accepted/deferred.
- The gating fix issue for auto-merge safety is complete before auto-merge is turned back on.

### PR DESCRIPTION
**Pull request by viamin**
## Summary

Remediate PRs in `viamin/agent-harness` that were merged before review completion, fixing forward 8 real defects discovered during audit of all flagged PRs from #91.

### Input validation
- **ProviderRuntime**: Added type validation for `model`, `base_url`, and `api_provider` attributes (must be String or nil)
- **ProviderRuntime**: Replaced `||` pattern in `.from_hash` with nil-aware `hash_value` helper to correctly handle falsy values
- **DockerCommandExecutor**: Reject whitespace-only `container_id` values

### Auth status consistency
- **Codex provider**: Ensured `:auth_method` key is present on all `auth_status` return paths
- **Gemini provider**: Ensured `:auth_method` key is present on all `auth_status` return paths

### Error pattern precision
- **Codex provider**: Added `\b` word boundaries to `/401/` regex in `error_patterns` to avoid false matches
- **Gemini provider**: Added `\b` word boundaries to `/429/` and `/503/` regexes in `error_patterns`

### Error handling
- **Kilocode provider**: Malformed version strings now rescued with a provider-specific error instead of bubbling a generic exception

### Documentation
- **AUDIT_DISPOSITION.md**: Full disposition record for all 8 audited PRs — categorizing 50+ original findings as already-addressed, 8 as fixed forward, and 2 as accepted non-defects

## Design Decisions

- **Fix forward over revert**: All defects were addressed with targeted fixes rather than reverting the original PRs, since the original changes are already in use and the defects are narrow in scope.
- **Disposition tracking**: Every audited PR received a recorded disposition per the issue's exit criteria, even when findings were already resolved during the original review cycle. This provides a clear paper trail for re-enabling auto-merge.
- **Word-boundary regex**: Error pattern regexes were tightened with `\b` anchors rather than restructured, keeping the fix minimal and reducing risk of breaking existing error classification.

---

Closes #91

### CODE REVIEW

**Comment by viamin** [comment]
## Agent Update

Commit succeeded. The CI lint failure was caused by markdownlint MD060 violations in `AUDIT_DISPOSITION.md` — the table separator rows used `|---------|-------------|` without spaces around pipes. Fixed by changing to `| ------- | ----------- |`. All checks pass:

- `bundle exec standardrb` — clean
- `markdownlint **/*.md --ignore CHANGELOG.md` — clean  
- `bundle exec rspec` — 1076 examples, 0 failures

File: lib/agent_harness/provider_runtime.rb [hunk 1]
```diff
@@ -97,13 +101,13 @@ def self.from_hash(hash)
...
-        flags: hash[:flags] || hash["flags"] || [],
-        unset_env: hash[:unset_env] || hash["unset_env"] || [],
-        metadata: hash[:metadata] || hash["metadata"] || {}
+        model: hash_value(hash, :model),
+        base_url: hash_value(hash, :base_url),
+        api_provider: hash_value(hash, :api_provider),
+        env: hash_value(hash, :env) || {},
+        flags: hash_value(hash, :flags) || [],
+        unset_env: hash_value(hash, :unset_env) || [],
+        metadata: hash_value(hash, :metadata) || {}
```
Comment (viamin) [inline lib/agent_harness/provider_runtime.rb]: The `hash_value` helper only fixes the scalar attributes here. These `|| {}` / `|| []` fallbacks still treat an explicit `false` the same as a missing key, so invalid input bypasses the initializer's type checks:

- `ProviderRuntime.from_hash(env: false)` currently returns `{}`
- `ProviderRuntime.from_hash(flags: false)` currently returns `[]`

That leaves the original bug partially unfixed for `env`, `flags`, `unset_env`, and `metadata`. Only `nil`/missing keys should default.

```suggestion        env: hash_value(hash, :env).nil? ? {} : hash_value(hash, :env),        flags: hash_value(hash, :flags).nil? ? [] : hash_value(hash, :flags),        unset_env: hash_value(hash, :unset_env).nil? ? [] : hash_value(hash, :unset_env),        metadata: hash_value(hash, :metadata).nil? ? {} : hash_value(hash, :metadata)```

**Review by viamin — COMMENTED** [review]
I found one actionable correctness issue in `ProviderRuntime.from_hash`: the new nil-aware helper fixes `false` for scalar fields, but the remaining `|| {}` / `|| []` fallbacks still collapse explicit `false` on structured fields to defaults instead of letting the initializer reject the invalid type.

**Comment by viamin** [comment]
## Agent Update

Commit succeeded. The fix addresses the reviewer's feedback: `from_hash` now uses `.nil?` checks instead of `||` for the collection attributes (`env`, `flags`, `unset_env`, `metadata`), so only `nil`/missing keys get defaults. An explicit `false` will now properly reach the initializer's type validation and raise an `ArgumentError` instead of being silently replaced with an empty collection.

- All 1076 tests pass
- Rubocop clean
- Pre-commit hooks passed

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Let me examine the full source files for a thorough review of the changes.

Now let me read the test files for full context:

Let me run the existing tests and lint first to establish a baseline:

All tests pass and lint is clean. Now let me do a thorough review of the diff to find issues.

I found several issues in the diff that need fixing:

1. **Codex `installation_contract` missing malformed version rescue** — Gemini and Kilocode both rescue `ArgumentError` from `Gem::Version.new`, but Codex doesn't. Passing a bad version bubbles a raw exception.
2. **`from_hash` calls `hash_value` twice per collection attribute** — Should use local variables.
3. **Missing tests** — No test for `from_hash` with falsy string attrs, no test for Codex malformed versions.
4. **`container_id` type safety** — Non-String values cause `NoMethodError` on `.strip` instead of clear `ArgumentError`.

Let me fix all of these:

Now fix Codex `installation_contract` to rescue malformed versions, consistent with Gemini and Kilocode:

Now add tests. First, add test for `from_hash` with falsy string attribute:

Add test for Codex malformed version:

Add test for DockerCommandExecutor with non-String container_id:

Now let me review the full diff of my changes to verify correctness:

Diff looks clean. Let me run lint and tests:

Lint clean, 1081 tests pass (5 new). Now let me commit:

Commit succeeded (pre-commit hooks passed — rubocop clean, all 1081 tests pass). Changes committed as `4354b91`.

**Summary of review fixes:**

- **`DockerCommandExecutor`**: Added `is_a?(String)` type check so non-String `container_id` values (e.g. Integer) raise `ArgumentError` instead of `NoMethodError` on `.strip`
- **`ProviderRuntime.from_hash`**: Eliminated redundant double `hash_value` calls for collection attributes by extracting to local variables
- **`Codex installation_contract`**: Added `rescue ArgumentError` around `Gem::Version.new` for malformed version strings, matching the pattern already applied to Gemini and Kilocode in this PR
- **5 new tests** covering falsy `from_hash` string attributes, Codex malformed versions, and non-String `container_id`

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Let me read the current state of all changed source files and their tests to do a thorough review.

Now let me read the test files to understand current test coverage:

Let me run lint and tests to verify the current state:

Let me check other providers for consistency with the patterns being applied:

Now let me check the base provider and authentication module for consistency issues that the reviewer might flag:

Let me check the current state of the repo and do a thorough review:

Let me do a thorough proactive scan. I want to check if `COMMON_ERROR_PATTERNS` in base.rb still has unbound numeric regexes, since Codex and Kilocode inherit from it:

The `COMMON_ERROR_PATTERNS` in `base.rb` has unbound numeric regexes (`/429/`, `/503/`, `/502/`) that Codex and Kilocode inherit directly, undermining the `\b` word boundary fix this PR applies to provider-specific patterns. I'll fix this at the source.

Now let me also check Anthropic's and Cursor's error patterns since they were flagged as having unbound numeric regexes:

Anthropic and Cursor have unbound numeric regexes too, but they define their own `error_patterns` entirely (not inheriting from `COMMON_ERROR_PATTERNS`) and are outside this PR's audit scope. The `base.rb` fix covers all providers that use the shared patterns.

Let me run lint and tests to verify:

Lint clean, all 1081 tests pass. Let me review the full diff for any remaining issues:

Let me check if there are existing tests for the error patterns to ensure completeness:

Good - the existing tests use `"HTTP 429"` and `"HTTP 503"` which still match with `\b` boundaries. All tests pass. Let me now do the final commit:

Committed as `bf036c4`. The fix adds `\b` word boundaries to the `/429/`, `/503/`, and `/502/` regexes in `COMMON_ERROR_PATTERNS` in `base.rb`. The previous commits only fixed these patterns in Codex and Gemini's provider-specific overrides, but Codex and Kilocode inherit the unbound patterns from the shared base. This completes the fix at the source, covering all providers that use `COMMON_ERROR_PATTERNS` (Codex, Kilocode, Aider, MistralVibe, OpenCode).

Pre-commit hooks passed: rubocop clean, all 1081 tests pass.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

I addressed the remaining actionable review gap and committed it as `199e47c` (`Review: fix generic status-code false positives`).

The issue was that the earlier `\b` regex hardening in provider patterns still got undermined by `ErrorTaxonomy`’s generic fallback, which was matching bare `429`/`401`/`503` substrings inside unrelated numbers. I fixed those generic status-code matches in [lib/agent_harness/error_taxonomy.rb](/workspace/lib/agent_harness/error_taxonomy.rb:122) and added regressions in [spec/agent_harness/error_taxonomy_spec.rb](/workspace/spec/agent_harness/error_taxonomy_spec.rb:44), [spec/agent_harness/providers/base_spec.rb](/workspace/spec/agent_harness/providers/base_spec.rb:104), plus one extra string-keyed falsy case in [spec/agent_harness/provider_runtime_spec.rb](/workspace/spec/agent_harness/provider_runtime_spec.rb:199).

Verification: `bundle exec rubocop` passed, `bundle exec rspec` passed with `1085 examples, 0 failures`, and the worktree is clean.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Let me read the source files in detail for the proactive scan.

Let me check the remaining providers for consistency.

Found one remaining consistency gap: Aider's `installation_contract` doesn't rescue `ArgumentError` from `Gem::Version.new`, same defect already fixed for Codex/Gemini/Kilocode. Fixing now.

Now add the test for Aider's malformed version handling:

Now run lint and tests:

Lint clean, 1086 tests pass (1 new). Let me verify the diff:

Diff looks clean. Committing:

Committed as `0e14995`. 

Found and fixed one remaining consistency gap: **Aider `installation_contract`** didn't rescue `ArgumentError` from `Gem::Version.new`, meaning a malformed version string like `"not-a-version"` would bubble a raw `ArgumentError: Malformed version number string...` instead of the provider-specific `"Unsupported Aider CLI version..."` message. This was the same defect pattern already fixed for Codex, Gemini, and Kilocode in earlier commits of this PR. Added a matching test. All 1086 tests pass, lint clean.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Let me install dependencies and verify the current state, then do a proactive scan of the full diff.

Lint clean, 1086 tests pass. Now let me do a thorough proactive scan of the full diff.

Let me read the current state of the files I need to modify.

Now I'll implement the fixes. Let me start with the nil/blank version guards and then add the missing tests.

Now add the missing tests:

The initializer still uses `||` for collections which swallows `false`. I need to fix the initializer to match the `from_hash` fix:

All 1093 tests pass, lint clean. Now let me commit.

Committed as `6db1695`. Pre-commit hooks passed (rubocop clean, all 1093 tests pass).

**Changes in this commit:**
- **Nil/blank version guards** in Aider, Codex, Kilocode `installation_contract` / `validate_install_version!` — prevents `Gem::Version.new(nil)` deprecation warning and `Gem::Version.new("")` silently producing version "0"
- **Initializer `.nil?` fix** — replaced `||` with `.nil?` ternary for `env`, `flags`, `metadata`, `unset_env` in `ProviderRuntime#initialize`, matching the `from_hash` fix so explicit `false` reaches type validation
- **7 new tests** — `version: nil` / `version: ""` for Aider/Codex/Kilocode, `from_hash(env: false)`, and Gemini expired-credentials `auth_method: :oauth` assertion

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

The bundle install failed. Let me check the Gemfile and fix this.

Dependencies installed. Now let me run lint and tests to establish a baseline.

All 1190 tests pass, lint clean. Now let me do a thorough proactive scan of the full diff. Let me read all the modified source files to look for any remaining issues.

Now let me check the remaining providers for consistency - specifically Anthropic, Cursor, GitHub Copilot, OpenCode, and MistralVibe for any similar gaps.

Found 4 issues across 3 providers. Let me read the files I need to fix.

Now let me apply the fixes. First, Gemini's missing nil/blank version guard:

Now fix Anthropic's bare numeric regexes:

Now fix Cursor's bare numeric regex:

Now I need to add tests. Let me check the existing test files for Gemini, Anthropic, and Cursor.

Let me read the relevant test sections to understand patterns and add tests.

Now let me add the tests. First, Gemini nil/empty version tests:

Now add Anthropic embedded numeric false-positive test:

Now add Cursor embedded numeric false-positive test:

Now run lint and tests:

There's a test failure. The Anthropic `error_patterns` has text patterns that may match. Let me debug which specific test case is failing:

The "task 50401 timeout" message matches the text pattern `/timeout/i`, not a numeric pattern. The word "timeout" is in the message string. I need to use a string that doesn't contain any of the text-based patterns. Let me fix that test case:

Wait, actually the point is to test that the numeric regexes don't false-positive. The "timeout" text pattern matching is expected behavior. The issue is that "50401" contains "504" but since I added `\b`, it shouldn't match. Let me verify the test string doesn't contain "timeout":

Now let me run the full suite:

All 1194 tests pass, lint clean. Now let me verify the full diff one more time before committing:

Clean diff. Now let me commit:

Committed as `ec169b0`. Pre-commit hooks passed (rubocop clean, all 1194 tests pass).

**Changes:**
- **Gemini `install_contract`**: Added nil/blank version guard before `Gem::Version.new`, consistent with Aider, Codex, Kilocode, and OpenCode
- **Anthropic `error_patterns`**: Added `\b` word boundaries to 7 bare numeric regexes (`/429/`, `/401/`, `/503/`, `/502/`, `/504/`, `/404/`, `/400/`)
- **Cursor `error_patterns`**: Added `\b` word boundary to bare `/429/` regex
- **8 new tests**: nil/empty version for Gemini, embedded numeric false-positive regression tests for Anthropic and Cursor

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Let me do a thorough scan of the remaining providers for consistency gaps.

Good - all 1194 tests pass, lint is clean. Now let me read the files that need fixing per the proactive scan.

Now let me fix both files. For Anthropic, I need to add a nil/blank guard and targeted rescue. For OpenCode, I need to refactor the rescue to only wrap `Gem::Version.new`.

Now let me find the test files and add appropriate tests.

Good - the OpenCode tests already cover nil/blank/malformed versions. Now let me add tests for Anthropic's nil/empty version handling.

Now let me do a thorough proactive scan of the full diff. Let me review the changes I just made and the overall diff for any remaining issues.

The changes look clean. Now let me run the full diff and scan for any remaining issues.

The diff looks clean. Now let me run lint and tests.

The `install_contract` method defaults nil to `SUPPORTED_CLI_VERSION` before calling `validate_version!`, so nil never reaches validation. Let me fix the test.

All 1196 tests pass, lint is clean. Now let me commit.

Committed as `a8953eb`. All 1196 tests pass, lint clean.

**Changes in this commit:**
- **Anthropic `validate_version!`**: Added nil/blank version guard and wrapped `Gem::Version.new` in targeted `begin/rescue ArgumentError`, matching the pattern used by all other providers. Previously relied on `version.to_s` which converted nil to `""` and produced unclear error messages.
- **OpenCode `normalize_install_version`**: Refactored to use targeted `begin/rescue` only around `Gem::Version.new`, matching the pattern in aider/codex/gemini/kilocode. The previous method-level `rescue ArgumentError` was over-broad and caught explicitly-raised errors too.
- **2 new tests** for Anthropic empty and whitespace-only version rejection.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Now let me look at the other providers to check for the same pattern and read the test files:

Let me read all the affected provider files:

Now I understand the full picture. Let me fix all affected providers. The pattern is: strip the version once at the top and use the normalized value throughout.

Now fix aider.rb — normalize version at the top:

Now fix codex.rb:

Now fix gemini.rb:

Now fix kilocode.rb:

Also fix kilocode's `validate_install_version!` to use `.empty?` instead of `.strip.empty?` since the version is already stripped:

Let me verify my edits are correct:

Good. Now let me find the test files for these providers to add tests:

Let me search for existing padded/whitespace version tests in the test files:

No existing tests for padded versions. Let me look at the test patterns for install_contract/installation_contract tests:

Now let me read the test files for aider, codex, gemini, and kilocode to understand their structure:

Now let me look at the end of the relevant test sections to add tests in the right place. Let me check each test file:

Good. Now I know where to add tests. Let me add tests for padded version strings. I'll add tests to all affected provider test files. First, let me see the exact context where I need to add tests:

Now add the test for Anthropic padded version normalization:

Now add tests for aider:

Now add tests for codex:

Now add tests for gemini:

Now add tests for kilocode:

There are 2 test failures for nil version handling. The `.strip` call raises `NoMethodError` on nil. Let me fix this:

For aider and codex, the default value is `SUPPORTED_CLI_VERSION` (a string), so nil only comes from explicit passing. The fix should handle nil before stripping:

Now fix gemini:

Now fix kilocode - the strip is in the caller (`installation_contract`), and nil goes through `validate_install_version!`:

Now for Anthropic, nil is the default and is handled by the `nil?` check:

Anthropic handles nil correctly via the ternary. Now let me run the tests:

One failure in the Gemini test. Let me check what package name Gemini uses:

The package is `@google/gemini-cli`, not `@anthropic-ai/gemini-cli`. Let me fix the test:

All 1202 tests pass, 0 failures. Now let me commit:

The diff looks clean. Now commit:

Missing racc gem. Let me install it:

The pre-commit hook runs `bundle exec rubocop` but the bundle environment doesn't have the right GEM_HOME. I need to add the gem bin to PATH:

Committed as `1fc1bcf`. The review comment identified that `validate_version!` strips whitespace before validation but `install_contract` used the untrimmed `target_version` for the install command and channel warning logic. Fixed by normalizing the version string (stripping whitespace) at the top of each provider's install contract method, before validation and before building commands/contracts. Applied the same fix across all 5 affected providers (Anthropic, Aider, Codex, Gemini, Kilocode) with 7 regression tests.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Only `anthropic.rb:36` is vulnerable. The fix is to add the same `respond_to?(:strip)` guard used by all other providers.

Wait, my edit has a precedence issue — the `if` modifier would prevent assignment when `version` is `nil` (since `nil` doesn't respond to `:strip`). Let me fix this properly, matching the pattern from other providers:

Now let me find and add a regression test. Let me look at the existing Anthropic tests:

Now I'll add a test for non-String version types (Symbol, Integer) after the whitespace-only test:

All 75 Anthropic tests pass, lint clean. Let me verify the new test specifically passes:

Now let me run the full test suite and commit:

All 1203 tests pass. Let me commit:

Committed as `393151d`. 

**Fix:** `install_contract` in `anthropic.rb:36` was the only provider calling `version.strip` without a type guard. Non-String inputs (e.g., `:latest`, `123`) raised `NoMethodError` before `validate_version!` could produce a clean `ArgumentError`. Fixed by splitting the assignment into two lines with a `respond_to?(:strip)` guard, matching the pattern already used by Aider, Codex, Gemini, and Kilocode. Added regression test for Symbol and Integer version inputs. All 1203 tests pass, lint clean.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

All 1203 tests pass, lint clean. Now let me review the full diff for any remaining issues.

Let me verify the `from_hash` handles all edge cases correctly with the current tests:

I found a subtle issue in kilocode.rb. When a non-String version (e.g., Integer) is passed, the error message shows `nil` instead of the original value because `normalized_version` gets nil from the conditional assignment. Let me fix this.

The issue: when `version` is non-String (e.g., `123`), `normalized_version` becomes `nil` via conditional assignment, and `validate_install_version!(nil)` produces an error message showing `nil` instead of `123`. All other providers (aider, codex, gemini) reassign `version` directly so the original value is preserved in error messages.

Now let me add a test to verify the error message preserves the original value for non-String inputs:

Now let me also verify the existing nil test still passes correctly (the error message should show `nil` for nil input, which is expected):

All kilocode tests pass. Now let me run the full suite and lint:

1204 tests pass, lint clean. Now let me commit:

Committed as `303fec9`.

**Fix:** `Kilocode.installation_contract` used a separate `normalized_version` variable with conditional assignment (`normalized_version = version.strip if version.respond_to?(:strip)`). When `version` was a non-String (e.g., `42`), Ruby initialized `normalized_version` to `nil`, producing error messages like "Unsupported Kilocode CLI version nil" instead of "Unsupported Kilocode CLI version 42". Fixed by reassigning `version` directly, matching the pattern used by aider, codex, and gemini. Added regression test. 1204 tests pass, lint clean.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
## Agent Update

Let me do a thorough proactive scan of the full diff, checking for any remaining issues.

Found 3 remaining issues from the proactive scan. Let me fix them all.

Now let me find the right places in the test files to add embedded numeric tests.

Gemini and Codex specs don't have these tests. Let me find the right insertion points.

Now let me add the missing tests:

Now add the 403/500 embedded numeric test to error_taxonomy_spec:

Lint clean, all 1206 tests pass. Now let me commit.

Committed as `3a5d32c`. Three fixes from the proactive scan:

- **`adapter.rb`**: Wrapped `Gem::Version.new(version)` in `rescue ArgumentError` in the fallback `install_command` path, matching the pattern already applied to all individual providers
- **`codex_spec.rb` / `gemini_spec.rb`**: Added embedded numeric substring regression tests for providers that had `\b`-bounded regexes but no corresponding tests
- **`error_taxonomy_spec.rb`**: Added non-matching test cases for HTTP 403 and 500 embedded in longer numbers

All 1206 tests pass, lint clean.

**Comment by viamin** [comment]
@codex review

**Comment by viamin** [comment]
**Escalation Note**

This has been escalated because the automated draft review limit (15 rounds) has been reached after 15 review cycles and the PR requires human intervention.
@viamin — manual review is required.

**How to resolve:**
- **Approve** this PR to allow auto-merge (if enabled)
- **Add the `paid-dismiss-escalation` label** to move back to the `ready` phase without merging
- **Convert to draft** on GitHub to restart the automated review cycle
